exonum_derive/src/tx_set.rs

```diff
@@ -0,0 +1,171 @@
+// Copyright 2018 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use proc_macro::TokenStream;
+use proc_macro2::{Ident, Span};
+use syn::{Data, DataEnum, DeriveInput, Type};
+
+struct TxSetVariant(u16, Ident, Type);
+
+fn get_tx_variants(data: &DataEnum) -> Vec<TxSetVariant> {
+    if data.variants.is_empty() {
+        panic!("TransactionSet enum should not be empty");
+    }
+
+    data.variants
+        .iter()
+        .enumerate()
+        .map(|(n, v)| {
+            if v.fields.iter().len() > 1 {
+                panic!("TransactionSet enum variant should have one field inside.");
+            }
+            let field = v
+                .fields
+                .iter()
+                .next()
+                .expect("TransactionSet enum variant can't be empty.");
+            TxSetVariant(n as u16, v.ident.clone(), field.ty.clone())
+        }).collect()
+}
+
+fn gen_conversions_for_transactions(
+    name: &Ident,
+    variants: &[TxSetVariant],
+    cr: &quote::ToTokens,
+) -> impl quote::ToTokens {
+    let conversions = variants.iter().map(|TxSetVariant(_, id, ty)| {
+        quote! {
+          impl Into<#name> for #ty {
```

Comment:
Why not `From<#ty> for #name`? Using `From` is preferable to `Into` if it's possible, as there is corresponding blanket `Into` implementation.

Reply:
I agree, this was copy-pasted  from `transactions!`. Maybe author used it for consistency with other conversions in `transactions!` which you can write in `Into` form only.